mlir/lib/Quantum/Transforms/VerifyParentGateAnalysis.hpp

```diff
@@ -0,0 +1,139 @@
+//        %0:2 = quantum.custom "CNOT"() %.., %..
+//        %1:2 = quantum.custom "CNOT"() %0#1, %0#0
+//     is not.
+//  3. If the gates are controlled, both gates' control wires and values
+//     must be the same. The control wires must be in the same order
+
+#pragma once
+
+#include "mlir/IR/BuiltinOps.h"
+#include "llvm/Support/Debug.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+
+template <typename OpType> class VerifyParentGateAnalysis {
+  public:
+    VerifyParentGateAnalysis(OpType gate)
+    {
+        ValueRange inQubits = gate.getInQubits();
+        auto parentGate = dyn_cast_or_null<OpType>(inQubits[0].getDefiningOp());
+
+        if (!verifyParentGateType(gate, parentGate)) {
+            verified = false;
+            return;
+        }
+
+        if (!verifyAllInQubits(gate, parentGate)) {
+            verified = false;
+            return;
+        }
+    }
+
+    bool getVerifierResult() { return verified; }
+
+  private:
+    bool verified = true;
+
+    bool verifyParentGateType(OpType op, OpType parentOp) const
+    {
+        // Verify that the parent gate is of the same type.
+        // If OpType is quantum.custom, also verify that parent gate has the
+        // same gate name.
+
+        if (!parentOp || !isa<OpType>(parentOp)) {
+            return false;
+        }
+
+        if (isa<quantum::CustomOp>(op)) {
+            StringRef opGateName = cast<quantum::CustomOp>(op).getGateName();
+            StringRef parentGateName = cast<quantum::CustomOp>(parentOp).getGateName();
+            if (opGateName != parentGateName) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    bool verifyAllInQubits(OpType op, OpType parentOp) const
+    {
+        // Verify that parent's results and current gate's inputs are in the same order
+        // If the gates are controlled, both gates' control wires and values
+        // must be the same. The control wires must be in the same order.
+
+        ValueRange inNonCtrlQubits = op.getNonCtrlQubitOperands();
+        ValueRange inCtrlQubits = op.getCtrlQubitOperands();
+        ValueRange parentOutNonCtrlQubits = parentOp.getNonCtrlQubitResults();
+        ValueRange parentOutCtrlQubits = parentOp.getCtrlQubitResults();
+
+        if ((inNonCtrlQubits.size() != parentOutNonCtrlQubits.size()) ||
+            (inCtrlQubits.size() != parentOutCtrlQubits.size())) {
+            return false;
+        }
+
+        for (const auto &[idx, qubit] : llvm::enumerate(inNonCtrlQubits)) {
+            if (qubit.getDefiningOp() != parentOp || qubit != parentOutNonCtrlQubits[idx]) {
+                return false;
+            }
+        }
+
+        for (const auto &[idx, qubit] : llvm::enumerate(inCtrlQubits)) {
+            if (qubit.getDefiningOp() != parentOp || qubit != parentOutCtrlQubits[idx]) {
```

Comment:
I think it is wrong for controls because for example Toffoli([0, 1, 2]) is the same as Toffoli([0, 2, 1]). So for controls we should compare the set.

Reply:
The three toffoli wires are technically non ctrl wires.

I think the control wires come more or less exclusively from [qml.ctrl](https://docs.pennylane.ai/en/stable/code/api/pennylane.ctrl.html), whose control wires are definitely not symmetric

Reply:
I meant `qml.ctrl(X(0), [1,2])` is equivalent to `qml.ctrl(X(0), [2, 1])`

Reply:
It will be solved in a different PR.